networks:
  traefik-net:
    external: true

services:
  traefik:
    image: traefik:v3.5.0
    container_name: traefik
    restart: always
    command: >
      sh -c "touch /letsencrypt/acme.json && chmod 600 /letsencrypt/acme.json && traefik 
      --log.level=INFO 
      --api.dashboard=false 
      --api=false 
      --api.insecure=false 
      --providers.docker=true 
      --providers.docker.exposedbydefault=false 
      --entrypoints.web.address=:80 
      --entrypoints.websecure.address=:443 
      --entrypoints.web.http.redirections.entrypoint.to=websecure 
      --entrypoints.web.http.redirections.entrypoint.scheme=https 
      --entrypoints.web.http.redirections.entrypoint.permanent=true 
      --certificatesresolvers.ssl-resolver.acme.tlschallenge=true 
      --certificatesresolvers.ssl-resolver.acme.email=desarrollo@indaws.es 
      --certificatesresolvers.ssl-resolver.acme.storage=/letsencrypt/acme.json"

    ports:
      - "80:80"
      - "443:443"
    networks:
      - traefik-net
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./letsencrypt:/letsencrypt
    labels:
      - "traefik.enable=true"

      # middlewares
      - "traefik.http.middlewares.gzip.compress=true"
    logging:
      driver: json-file
      options:
        max-size: 10m

  # error page
  errorpage:
    image: nginx:alpine
    container_name: traefik-errorpage
    restart: always
    volumes:
      - ./errorpage:/usr/share/nginx/html:ro
      - ./nginx-error.conf:/etc/nginx/conf.d/default.conf:ro
    networks:
      - traefik-net
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik-net"

      # maintenance page
      - "traefik.http.services.maintenance-service.loadbalancer.server.port=80"
      - "traefik.http.services.maintenance-service.loadbalancer.server.scheme=http"
      - "traefik.http.services.maintenance-service.loadbalancer.passhostheader=false"

      # 403 service
      - "traefik.http.services.forbidden-service.loadbalancer.server.port=80"
      - "traefik.http.services.forbidden-service.loadbalancer.server.scheme=http"
      - "traefik.http.services.forbidden-service.loadbalancer.passhostheader=false"

      # /web/database/* - priority very high
      - "traefik.http.routers.database-blocked.rule=(PathPrefix(`/web/database/`) || PathPrefix(`/asoo/auth/`) || PathPrefix(`/asoo/web/dataset/`)) && !ClientIP(`194.30.4.97`) && !ClientIP(`194.30.4.101`) && !ClientIP(`172.16.100.1`) && !ClientIP(`10.10.10.1`)"
      - "traefik.http.routers.database-blocked.entrypoints=websecure"
      - "traefik.http.routers.database-blocked.tls=true"
      - "traefik.http.routers.database-blocked.service=forbidden-service"
      - "traefik.http.routers.database-blocked.middlewares=add-forbidden-header"
      - "traefik.http.routers.database-blocked.priority=1000"

      # known domains - priority very low
      - "traefik.http.routers.errorpage-fallback.rule=PathPrefix(`/`)"
      - "traefik.http.routers.errorpage-fallback.entrypoints=websecure"
      - "traefik.http.routers.errorpage-fallback.tls=true"
      - "traefik.http.routers.errorpage-fallback.service=maintenance-service"
      - "traefik.http.routers.errorpage-fallback.priority=-1"

      # 403 middleware
      - "traefik.http.middlewares.add-forbidden-header.headers.customrequestheaders.X-Forbidden-Access=true"
    logging:
      driver: json-file
      options:
        max-size: 10m
